packages/tabs/src/Tabs/Tabs.tsx

```diff
@@ -74,62 +67,81 @@ const Tabs = (props: AccessibleTabsProps) => {
   const { theme, darkMode } = useDarkMode(darkModeProp);
-  const accessibilityProps = {
-    ['aria-label']: ariaLabel,
-    ['aria-labelledby']: ariaLabelledby,
-  };
-
   const id = useIdAllocator({ prefix: rest.id || 'tabs' });
+
   const { descendants: tabDescendants, dispatch: tabDispatch } =
     useInitDescendants<HTMLDivElement>();
   const { descendants: tabPanelDescendants, dispatch: tabPanelDispatch } =
     useInitDescendants<HTMLDivElement>();
 
+  const isControlled = typeof controlledSelected !== 'undefined';
   const [uncontrolledSelected, setUncontrolledSelected] = useState(0);
   const [selected, setSelected] = [
-    controlledSelected ? controlledSelected : uncontrolledSelected,
-    controlledSelected ? setControlledSelected : setUncontrolledSelected,
+    isControlled ? controlledSelected : uncontrolledSelected,
+    isControlled ? setControlledSelected : setUncontrolledSelected,
   ];
 
-  const childrenArray = useMemo(
-    () => React.Children.toArray(children) as Array<React.ReactElement>,
-    [children],
-  );
+  const accessibilityProps = {
+    ['aria-label']: ariaLabel,
+    ['aria-labelledby']: ariaLabelledby,
+  };
 
-  const handleChange = useCallback(
+  const handleClickTab = useCallback(
     (e: React.SyntheticEvent<Element, MouseEvent>, index: number) => {
       setSelected?.(index);
     },
     [setSelected],
   );
 
-  const getEnabledIndexes: () => [Array<number>, number] = useCallback(() => {
-    const enabledIndexes = childrenArray
-      .filter(child => !child.props.disabled)
-      .map(child => childrenArray.indexOf(child));
-
-    return [enabledIndexes, enabledIndexes.indexOf(selected!)];
-  }, [childrenArray, selected]);
-
-  const handleArrowKeyPress = useCallback(
+  const tabTitleElements = tabDescendants.map(
+    descendant => descendant.element.parentNode as HTMLElement,
+  );
+  const getActiveAndEnabledIndices: () => {
+    activeIndex: number;
+    enabledIndices: Array<number>;
+  } = useCallback(() => {
+    const enabledIndices = tabTitleElements
+      .filter(tabTitleEl => !tabTitleEl.hasAttribute('disabled'))
+      .map(tabTitleEl => tabTitleElements.indexOf(tabTitleEl));
+
+    return {
+      activeIndex: enabledIndices.indexOf(selected),
+      enabledIndices,
+    };
+  }, [selected, tabTitleElements]);
+
+  const handleKeyDown = useCallback(
     (e: KeyboardEvent) => {
-      if (!(e.metaKey || e.ctrlKey)) {
-        if (e.key === keyMap.ArrowRight) {
-          const [enabledIndexes, current] = getEnabledIndexes();
-          setSelected?.(enabledIndexes[(current + 1) % enabledIndexes.length]);
-        } else if (e.key === keyMap.ArrowLeft) {
-          const [enabledIndexes, current] = getEnabledIndexes();
-          setSelected?.(
-            enabledIndexes[
-              (current - 1 + enabledIndexes.length) % enabledIndexes.length
-            ],
-          );
-        }
-      }
+      if (e.metaKey || e.ctrlKey) return;
```

Comment:
should `shift` technically move in the opposite direction?

Reply:
hmm good question. I don't see that pattern adopted with tabs implemented by [MUI](https://mui.com/material-ui/react-tabs/) and [react aria](https://react-spectrum.adobe.com/react-aria/examples/category-tabs.html). My guess is that because arrow keys have an implied direction, it's superfluous to create additional `shift` + arrow key combinations